docs: Add token-2022 audits and clarify upgradability (#3894)
solana-labs/solana-program-library: docs/src/token-2022.md

The Token-2022 Program is currently under multiple audits to ensure safety of
funds. All audits will be published here as they are completed.

> Here are the completed audits of 12 December 2022:

Reviewer: ```suggestion
Here are the completed audits as of 12 December 2022:
```

* Halborn
    - Review commit hash [`c3137a`](https://github.com/solana-labs/solana-program-library/tree/c3137af9dfa2cc0873cc84c4418dea88ac542965/token/program-2022)
    - Final report https://github.com/solana-labs/security-audits/blob/master/spl/HalbornToken2022Audit.pdf
* Zellic
    - Review commit hash [`54695b`](https://github.com/solana-labs/solana-program-library/tree/54695b233484722458b18c0e26ebb8334f98422c/token/program-2022)
    - Final report https://github.com/solana-labs/security-audits/blob/master/spl/ZellicToken2022Audit.pdf

## Status and Upgradability

> The Token-2022 program is still under audit, so it is considered beta software,
> and not meant for full production use. All clusters have the program for testing,

Reviewer 2: ```suggestion
The Token-2022 program is still under audit and not meant for full production use. All clusters have the program for testing,
```

“Mainnet Beta” :upside_down_face:

Author: It's beta software all the way down!

but please do not mint production tokens with it yet.

> To facilitate deploying updates and security fixes, the program is still upgradable,

Reviewer 2: Do we want to block finalization on 2xtx too? I feel yes, in the event that we discover a fail flaw with confidential transfers at the very end of 2xtx development.

Author: That might be overly cautious.  Confidential transfers have been considered and fixed on every audit, so it should be considered safe. With that said, yeah, we can probably wait

> but the program will be frozen and marked unupgradable once all audits are complete.

Author: We're in the process of getting them finalized for publishing, so I'm guessing it should be fine. I'll double check

> The Token-2022 program is still under audit, so it is considered beta software,
> and not meant for full production use. All clusters have the program for testing,
> but please do not mint production tokens with it yet.
> 
> To facilitate deploying updates and security fixes, the program is still upgradable,
> but the program will be frozen and marked unupgradable once all audits are complete.
> This will likely happen sometime in Q1 2023.

Reviewer: ```suggestion
The Token-2022 program is still under audit. It should be considered beta software
and not ready for full production use. All clusters have the program deployed **for
testing and development purposes ONLY**.

To facilitate deploying updates and security fixes, the program deployment remains
upgradeable. Once audits are complete, the deployment will be marked final and no
further upgrades will be possible. This is expected to happen sometime in Q1 2023.
```
